### PR TITLE
github - shellcheck workflow

### LINKED ISSUE #3489: Add a github workflow to run shellcheck on shellscripts
Add shellcheck to the builder image https://github.com/build-trust/ockam/blob/develop/tools/docker/builder/Dockerfile

Add a new workflow here https://github.com/build-trust/ockam/tree/develop/.github/workflows
that will run whenever any shellscript in the repo is added or changed

### PR DESCRIPTION
**Pull request by elpiel**
Resolves #3489 

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. - https://github.com/build-trust/ockam-contributors/pull/134

### CODE REVIEW

File: .github/workflows/shellcheck.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+jobs:
+  lint:
+    name: Shellcheck - Lint
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check out ockam repository
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+      - name: Run ShellCheck
+        uses: ludeeus/action-shellcheck@master
```
Comment (metaclips) [inline .github/workflows/shellcheck.yml]: Let's pin this workflow to latest master SHA `b2bbefc2e6b9dcbc6355b85e366c9e55bf8d57e1`
```
uses: ludeeus/action-shellcheck@b2bbefc2e6b9dcbc6355b85e366c9e55bf8d57e1
```

**Review by metaclips — CHANGES_REQUESTED** [review]
Thanks @elpiel 🚀

File: .github/workflows/shellcheck.yml (hunk 1 above)
Comment (elpiel) [reply .github/workflows/shellcheck.yml]: Fixed.

**Comment by elpiel** [comment]
Great! Could you run the CI to see if the workflow is good?

**Comment by mrinalwadhwa** [comment]
@elpiel left a small request for you [here](https://github.com/build-trust/ockam-contributors/pull/134). This will satisfy the commits linter workflow.

**Comment by metaclips-botter** [comment]
☑️ Orchestrator test on PR #3689 at commit `1407fba7ef5da87835769ebe390599cd6fbf375f` passed

**Comment by elpiel** [comment]
Keep in mind that this PR is made from a fork of Lechev.space. This is why I'm planning to also add Lechev.space as a Contributor

**Comment by mrinalwadhwa** [comment]
for some reason, actions isn't triggering the new workflow in this PR 🤔

**Comment by elpiel** [comment]
Right now, after pinning, it gives me this error:
Screenshot from 2022-10-21 09-43-04

**Comment by mrinalwadhwa** [comment]
Ah I see because it uses a new third party action. Would you mind changing it use the official shell check container image so we can avoid an extra set of dependencies?

**Comment by elpiel** [comment]
Ok, let's try this then. I'm unsure if `.` as a parameter will work for the whole folder.

**Comment by elpiel** [comment]
Even though there is a new entrypoint supplied for the container, GH overrides it and uses tail (which does not exist in the image apparently)

```
/usr/bin/docker create --name 2ce7873918834ed88b0ee098a4280d20_koalamanshellchecksha256722015803bb3606fd163c475cf4141ef8f39b6485da5ee0cbd8611042396b7da_6f325f --label 8d5581 --workdir /__w/ockam/ockam --network github_network_af4c9cd240cb4496a1ff47f348f87089 --entrypoint sh -e "HOME=/github/home" -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work":"/__w" -v "/home/runner/runners/2.298.2/externals":"/__e":ro -v "/home/runner/work/_temp":"/__w/_temp" -v "/home/runner/work/_actions":"/__w/_actions" -v "/opt/hostedtoolcache":"/__t" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" --entrypoint "tail" koalaman/shellcheck@sha256:722015803bb3606fd163c475cf4141ef8f39b6485da5ee0cbd8611042396b7da "-f" "/dev/null"
```

Both entrypoints are visible, but since the `tail` is last, it's the one taken for the command.
GH workflow documentation does not mention anything about not supporting this option.

How should we proceed?

**Comment by mrinalwadhwa** [comment]
Hmm, looks like others have faced the same problem
https://github.com/actions/runner/issues/1964

A couple of choices:

1. shellcheck is preinstalled, just use that:
https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md

2. try the `koalaman/shellcheck-alpine` image mentioned in the shellcheck readme 
https://github.com/koalaman/shellcheck#installing

3. apt install on download an exact version of shellcheck

I think option 1 may be good enough.

File: .github/workflows/shellcheck.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+jobs:
+  lint:
+    name: Shellcheck - Lint
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Check out ockam repository
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+      - name: Run ShellCheck
+        run: shellcheck .
```
Comment (metaclips) [inline .github/workflows/shellcheck.yml]: We should run shellcheck on shell scripts across all subdirectories
```suggestion
        run: shellcheck **/*.sh
```

**Review by metaclips — CHANGES_REQUESTED** [review]
Little change for a merge

File: .github/workflows/shellcheck.yml (hunk 2 above)
Comment (elpiel) [reply .github/workflows/shellcheck.yml]: Good point. I wasn't sure if `.` will just work for the whole folder and subfolders

Comment (elpiel) [reply .github/workflows/shellcheck.yml]: Ahh... Does not seem to work though. Should the globbing be `**/**.sh` or maybe `.` just works? 🤔
I haven't really tried it locally

**Comment by elpiel** [comment]
@metaclips should be good now. I had to use `find` as `shellcheck` doesn't support globbing

**Review by mrinalwadhwa — APPROVED** [review]

**Comment by mrinalwadhwa** [comment]
Merged in https://github.com/build-trust/ockam/pull/3752

@elpiel Thank you for an awesome contribution 🥳

**Comment by mrinalwadhwa** [comment]
I also added a followup issue to fix all shellcheck errors https://github.com/build-trust/ockam/issues/3754

**Pull request closed without merging**